[tune] Restructure trial directory

ujvl opened the pull request:
Restructures trial and trainable directory to avoid complex syncing scenarios, eg: race condition between rsync of logs and trainable writing checkpoint. Partially addresses #6226

Previously the directory structure was flat, so we had something like (on both the driver and worker):
```
MyTrainableName_asdfasdf_123/     <-- logdir
  progress.csv
  params.json
  checkpoint1/
  checkpoint2/
  trainable.log1
  trainable.log2
```
It is now structured as:
```
MyTrainableName_asdfasdf_123/     <-- trial_dir (logdir will continue pointing here for backwards-compat)
  progress.csv
  params.json
  checkpoints/                    <-- checkpoint_dir
    checkpoint1/
    checkpoint2/
  remote_logs/                    <-- remote_logdir
    trainable.log1
    trainable.log2
``` 

Logs written by the trainable are now isolated in `remote_logdir`. 
Checkpoints are now isolated in `checkpoints`. 

This lets us _selectively_ sync the directories we want. ie, on result we now _only_ sync-down `remote_logdir`. On checkpoint we only sync-down the `checkpoint_dir`. This prevents log syncing (which is non-blocking) to conflict with checkpoints that are simultaneously being written at the worker.

This PR also:
- refactors the confusing NodeSyncMixin code and 
- allows subdirectory syncs

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://ray.readthedocs.io/en/latest/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/18979/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/18982/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/18999/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/19003/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/19009/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/19011/
Test FAILed.

richardliaw commented on python/ray/tune/sync_client.py:
```diff
@@ -0,0 +1,141 @@
+from __future__ import absolute_import
```
this is just a move right?

ujvl replied:
yeah idt i changed anything

ujvl replied:
also moved `log_sync.py` contents into `syncer.py` but there are some minor changes there (eg having `NodeSyncer` extend `LogSyncer` instead of using mixin

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/19013/
Test FAILed.

richardliaw commented on python/ray/tune/trial.py:
```diff
@@ -48,6 +50,53 @@ def __str__(self):
             return "{}:{}".format(self.hostname, self.pid)
 
 
+class TrialDirectory(object):
```
Can you rename this? Or else I'm getting something confusing like:

trial.trial_dir.logdir
trial.logdir
trial.trial_dir


and it's not quite clear which is an object, which is a string, which is a path, etc.

richardliaw commented on python/ray/tune/trial.py:
```diff
@@ -48,6 +50,53 @@ def __str__(self):
+    REMOTE_LOGDIR = "remote_logs"  # logs that originated on the remote worker
+    CHECKPOINT_DIR = "checkpoints"  # checkpoints on the remote worker
+
+    def __init__(self, identifier, local_dir):
```
I think this should be `trial_name`?

richardliaw commented on python/ray/tune/trial.py:
```diff
@@ -189,9 +240,25 @@ def __init__(self,
         if trial_name_creator:
             self.custom_trial_name = trial_name_creator(self)
 
+    def init(self):
```
the nice thing about making this specifically `init_logger` was that in various places of the codebase, we stop the trial but keep the logger. Changing this to `init` makes it harder to think about (it's asymmetric to "stop").

ujvl replied:
ah gotcha, will refactor

richardliaw commented:
And also, does this break backwards compat for users who have built tools that specifically depend on a particular substructure of the `local_dir`, specifically for tfevents and csvs?

ujvl commented:
> At a high level, can you provide some context for what happens before and after?
updated the PR description w/ some context
> And also, does this break backwards compat for users who have built tools that specifically depend on a particular substructure of the local_dir, specifically for tfevents and csvs?

earlier i changed `logdir` to be `trial_dir/driver_logdir`, but that isn't backwards compatible so i just reverted back to `trial_dir=logdir` (no change), that should work. checkpoints and trainable logs still selectively synced.

ujvl commented:
jenkins test tune

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Tune-Tests/308/
Tune tests failed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/19016/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/19020/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/19021/
Test FAILed.

richardliaw commented on python/ray/tune/track/session.py:
```diff
@@ -70,8 +70,8 @@ def _initialize_logging(self,
 
-        # TODO(rliaw): Refactor `logdir` to `trial_dir`.
-        self._logdir = Trial.create_logdir(trial_name, self._experiment_dir)
+        self._trial_dir = TrialDirSchema(trial_name, self._experiment_dir)
```
_trial_dir_schema?

richardliaw commented on python/ray/tune/trial.py:
```diff
@@ -162,15 +219,21 @@ def __init__(self,
 
+        self.trial_dir_schema = TrialDirSchema(str(self), self.local_dir)
+        self.logdir = self.trial_dir_schema.logdir
+        self.trial_dir_schema.mkdir()
```
mkdir is automatic right?

richardliaw commented on python/ray/tune/trainable.py:
```diff
@@ -74,14 +74,15 @@ def __init__(self, config=None, logger_creator=None):
-                prefix=logdir_prefix, dir=DEFAULT_RESULTS_DIR)
+            name = self.__class__.__name__
+            dir_schema = TrialDirSchema(name, DEFAULT_RESULTS_DIR)
+            dir_schema.mkdir()
```
no need for mkdir?

richardliaw replied:
Or you can take it out of the __init__

richardliaw commented on python/ray/tune/ray_trial_executor.py:
```diff
@@ -108,19 +108,21 @@ def _setup_remote_runner(self, trial, reuse_allowed):
-        return cls.remote(config=trial.config, logger_creator=logger_creator)
+        runner = cls.remote(config=trial.config, logger_creator=logger_creator)
+        # Initialize the checkpoint directory.
+        runner._init_checkpoint_dir.remote(trial.checkpoint_dir)
```
Why not just move this into logger_creator? 

After all, the trial_dir_schema is already in the logger creator.

richardliaw commented on python/ray/tune/trial.py:
```diff
@@ -238,18 +299,23 @@ def update_resources(self, cpu, gpu, **kwargs):
-        self.address = location
+        self.location = location
+        if self.syncer:
+            self.syncer.set_worker_ip(self.location.hostname)
```
why does the result_logger not need to be set here?

richardliaw commented on python/ray/tune/trial.py:
```diff
@@ -162,15 +219,21 @@ def __init__(self,
+        self.trial_dir_schema = TrialDirSchema(str(self), self.local_dir)
+        self.logdir = self.trial_dir_schema.logdir
+        self.trial_dir_schema.mkdir()
+        self.syncer = get_syncer(self.trial_dir, self.trial_dir,
```
Why not create this when the result_logger is created? 

Can you also rename this to `checkpoint_syncer`? Otherwise, it's hard to immediately know the difference.

ujvl replied:
keeping as syncer since it syncs results too

richardliaw commented on python/ray/tune/trial.py:
```diff
@@ -310,13 +376,11 @@ def on_checkpoint(self, checkpoint):
         """
         if self.sync_on_checkpoint and checkpoint.storage == Checkpoint.DISK:
-            # Wait for any other syncs to finish. We need to sync again after
-            # this to handle checkpoints taken mid-sync.
```
can you keep these comments?

richardliaw commented on python/ray/tune/trial.py:
```diff
@@ -346,6 +410,7 @@ def update_last_result(self, result, terminate=False):
         self.last_result = result
         self.last_update_time = time.time()
         self.result_logger.on_result(self.last_result)
+        self.syncer.sync_down(TrialDirSchema.REMOTE_LOGDIR)
```
wait ... the syncer is in charge of both results and checkpoints?

Then why is there a separate syncer in the UnifiedLogger?
And also, why specify the source and target in instantiation in the first place then?

ujvl replied:
> Then why is there a separate syncer in the UnifiedLogger?

so actually there's only one syncer. I'm passing in a no-op to `UnifiedLogger`. it makes more sense to me to have the trial control the syncing (since it's aware of the directory structure and logging should be separate from checkpointing, conceptually).

> And also, why specify the source and target in instantiation in the first place then?

yeah you're right i just did it this way since it was less code to change, i can change it to take the full paths at execution.

richardliaw replied:
Then we don't need a syncer in the UnifiedLogger anymore right?

ujvl replied:
yeah i left it in since i wasn't sure if other places were using that syncing functionality but it seems like they don't so i just got rid of it

richardliaw commented on python/ray/tune/tests/test_tune_save_restore.py:
```diff
@@ -50,57 +53,50 @@ def tearDown(self):
             },
             checkpoint_freq=1,
-            local_dir=local_dir,
+            local_dir=absolute_local_dir,
```
I think the point of some of these tests was to specifically test if you could pass in a relative local dir in addition to an absolute one...

richardliaw commented on python/ray/tune/sync_client.py:
```diff
@@ -0,0 +1,141 @@
+            raise TuneError("{} is not a string.".format(sync_string))
+        if "{source}" not in sync_string:
+            raise TuneError("Sync template missing '{source}'.")
+        if "{target}" not in sync_string:
```
why change this?

richardliaw replied:
these were originally ValueErrors

ujvl replied:
whoops was trying to get tests to pass and forgot to revert

richardliaw commented on python/ray/tune/ray_trial_executor.py:
```diff
@@ -108,19 +108,21 @@ def _setup_remote_runner(self, trial, reuse_allowed):
-            if not os.path.exists(remote_logdir):
-                os.makedirs(remote_logdir)
+            trial_dir_schema.mkdir()
             if not ray.worker._mode() == ray.worker.LOCAL_MODE:
```
The local mode is tricky; there is a PR/issue here (https://github.com/ray-project/ray/issues/6228) to track this - I think we should maybe just move this all to a LocalModeExecutor.

richardliaw commented:
Thanks for the fast review!

> I remember at some point we discussed refactoring the logging/syncing such that not everything has to go through the driver and for example checkpoints could be pushed directly to S3. Is there any plan for this?

Yes, I'll (or Ujval) probably work on this right after this is merged.

richardliaw commented on python/ray/tune/trial.py:
```diff
@@ -48,6 +50,61 @@ def __str__(self):
+    def logdir(self):
+        """Directory containing logs that originated on the driver.
+
+        For backwards compatibility, this is set to the root.
```
So the backwards compat is needed because users have weird scripts written to navigate and parse the outputs assuming this particular directory structure. 

Now if that wasn't an issue, I'm not sure if there's any difference in putting the driver logs in a specific folder, but maybe I'm wrong? @ujvl.

ujvl replied:
yeah there's no real benefit from placing it in a subdir, aside from clarity

ujvl commented on python/ray/tune/syncer.py:
```diff
@@ -33,127 +34,35 @@ def wait_for_sync():
+        return None
 
-NOOP = FunctionBasedClient(lambda s, t: None, lambda s, t: None)
+    rsh = "ssh -i {ssh_key} -o ConnectTimeout=120s -o StrictHostKeyChecking=no"
```
remote shell i believe (`-e` == `--rsh`)

ujvl commented on python/ray/tune/trial.py:
```diff
@@ -48,6 +50,61 @@ def __str__(self):
+
+    def __init__(self, trial_name, local_dir):
+        local_dir = os.path.expanduser(local_dir)
+        if not os.path.exists(local_dir):
```
not python2.7 compatible unfortunately

ujvl commented:
@richardliaw 

> I'm also thinking about the usage between using a single Trainable and tune.run. i.e., in the RLlib case

maybe it makes sense to rename `remote_logs` to `trainable_logs` and then have RLLib trainables write there as well? then transition is seamless

>  I remember at some point we discussed refactoring the logging/syncing such that not everything has to go through the driver and for example checkpoints could be pushed directly to S3. Is there any plan for this?

yeah @hartikainen i'll put out a PR for this some time in the next couple weeks perhaps

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/19059/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/19062/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/19068/
Test FAILed.

richardliaw commented:
Actually, let's keep the RLlib case as close as possible to the previous behavior and then document it. I think this means keeping it as:

```
trial_dir/
    custom.log
    custom.log2
    checkpoint_dir/
       ...
```

> btw, i'm unclear on how custom logging works: since tune always passes in a no-op logger the only way for users to actually use their own logger is override __init__. Does RLLib do that and ignore the logger passed in? because if not I don't see why we're syncing anything in the first place.

`Trainable.remote()` creates a process. Inside the process, there is a `os.chdir`. This allows users to write directly to disk from inside the process. This should be synced back to the driver. This does not show up in the rllib-only case. 

>> Checkpoints are Tune only, so non-Tune usage doesn't need to create a dir. save takes in a checkpoint_dir.

> can't a user call my_trainable.save()? In fact rllib tests do this.

Yeah, that's right.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/19098/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/19108/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/19148/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/19161/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/19164/
Test FAILed.

ujvl commented:
jenkins test tune

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Tune-Tests/312/
Tune tests failed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/19220/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/19221/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/19230/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/19235/
Test FAILed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/19259/
Test PASSed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/19262/
Test FAILed.

ujvl commented:
Closing in favor of #6376 which borrows some elements of this PR.
